### ISSUE TITLE
Add a helper function for getting the body of an HTTP request.

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6251]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)| @tomprince reported|
|-|-|
|Trac ID|trac#6251|
|Type|enhancement|
|Created|2013-01-10 22:04:09Z|
|Branch|https://github.com/twisted/twisted/tree/http-getBody-6251-2|

Very often, people just want the body of a request, and aren't going to handle it incrementally. Rather than making everyone write their own protocol, just provide a helper function to do that.

Attachments:

* [getBody.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/a66/a66ad20c365378cb21fb86cec19b64bde54fb87c/13b48b836ed6c4b766eac68c5b9614c6b1dae43f.patch) (3932 bytes) - added by tom.prince on 2013-01-10 22:05:02Z - Initial patch (lacks tests)

<details><summary>Searchable metadata</summary>

```
trac-id__6251 6251
type__enhancement enhancement
reporter__tom_prince tom.prince
priority__normal normal
milestone__ 
branch__branches_http_getBody_6251_2 branches/http-getBody-6251-2
branch_author__exarkun__tomprince exarkun, tomprince
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1357855449000000 1357855449000000
changetime__1370057122000000 1370057122000000
version__None None
owner__exarkun exarkun
cc__jknight cc__Julian
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_2"></a>|@tomprince commented|
|-|-|

To expand on the description, the original client api was ```twisted.web.client.getPage```, which just returned the body of the response. But that api is insufficient if you want any more information about the response.

The newer API is ```twisted.web.client.Agent```, which returns an ```IResponse``` object, that has all the relevant information about the response. To get the body of the response, you need to pass a ```IProtocol``` object to ```IResponse.deliverBody```, which will deliver the body incrementally to the protocol, as the response is recieved.

If the user isn't going to process the response body incrementally, they currently need to implement a protocol that buffers the body until it is complete. Rather than making everybody reimplement this logic, as function should be provided to call ```.deliverBody``` with an appropriate protocol and return a deferred that fires with the body, once it has been received.

**Comment by twisted-trac** [comment]
|[<img alt="Julian's avatar" src="https://avatars.githubusercontent.com/u/329822?s=50" width="50" height="50">](https://github.com/Julian)<a name="note_3"></a>|@Julian commented|
|-|-|

I like this. There's some code to do this collection in treq but I also think it'd be nice if Agent could do so already.

Think it'd be better to move the accumulator out of proto_helpers and into the client module if that patch is going to be built upon.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

> Think it'd be better to move the accumulator out of proto_helpers and into the client module

Not just better, of course.  Completely necessary.  Implementation code may never import code from a test module.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_5"></a>|@tomprince commented|
|-|-|

(In [36797]) Branching to http-getBody-6251

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_6"></a>|@tomprince set owner to @tomprince|
|-|-|

[treq's implementation](https://github.com/dreid/treq/blob/master/treq/content.py)
* It uses a weak key dictionary to cache the content of a given response.
* It also has helpers to parse json, and decode to unicode (respecting content-type headers).
* It has a helper that injects a provided function as the `dataReceived` of a protocol, and returns a deferred that fires when the body is complete.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)|@tomprince removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve set owner to @tomprince|
|-|-|

1. Relevant twistedchecker issues:
```
W9016:2097,0: Too many blank lines after docstring, found 2
W9208:2109,8:GetBodyTests.FakeResponse.deliverBody: Missing docstring
W9208:2113,4:GetBodyTests.test_simple: Missing docstring
W9208:2122,4:GetBodyTests.test_withPotentialDataLoss: Missing docstring
W9208:2140,4:GetBodyTests.test_withConnectionLost: Missing docstring
W9012:2173,0: Expected 2 blank lines, found 3
W9012:2190,0: Expected 2 blank lines, found 4
```

2. _GetBodyProtocol checks ConnectionDone in connectionLost, but it should check ResponseDone instead.

3. In the example, you need to return d from the main function for reactor to work.

4. There is a typo in the example: 'Rsponse body:'

5. "buf" is not a great variable name. "dataBuffer" maybe?

6.
```
+    This is a helper function, for clients that don't want to incrementally                                       
+    recieve the body of an HTTP response.
```
Typo: recieve.

Thanks, please put it back for review once fixed.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

> +    This is a helper function, for clients that don't want to incrementally

That comma shouldn't be there, either. ;)  A good rule of thumb for commas is "when in doubt leave it out".

**Comment by twisted-trac** [comment]
|[<img alt="Julian's avatar" src="https://avatars.githubusercontent.com/u/329822?s=50" width="50" height="50">](https://github.com/Julian)<a name="note_10"></a>|@Julian commented|
|-|-|

`getJSONBody` is not really that useful as written. I presume that it was meant to use the encoding stuff that was added (`getTextBody`) rather than `getBody` directly?

For reference, requests does: https://github.com/kennethreitz/requests/blob/master/requests/models.py#L630 as per the JSON RPC for guessing encodings, which is useful if we're adding something JSON specific, but wouldn't devastate me personally if it only was present in treq, not in Twisted itself, as long as `getBody` was there.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

(In [38566]) Branching to 'http-getBody-6251-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

I took the JSON and text parsing stuff out.  Those should come back as part of a follow-up ticket.  I also renamed the function to `readBody`, sorry.


Build results - http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/http-getBody-6251-2

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_15"></a>|@therve set owner to @exarkun|
|-|-|

1. There are 2 twistedchecker errors: http://buildbot.twistedmatrix.com/builders/twistedchecker/builds/722/steps/run-twistedchecker/logs/new%20twistedchecker%20errors

2. The howto doc should be formatted to fit in 79 columns.

3. The reference to example says "Inspect the response", which is the same as the previous example. It should probably mention "and display the body".

Thanks, please merge once fixed!

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_16"></a>|@tomprince commented|
|-|-|

>  2. The howto doc should be formatted to fit in 79 columns. 

There has been talk about switching to sentence per line, for docs. Doing this provides more sensible diffs, rather than having to rewrap each time.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun commented|
|-|-|

(In [38573]) Address review comments

refs [#6251](https://github.com/twisted/twisted/issues/6251)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun commented|
|-|-|

> There has been talk about switching to sentence per line, for docs. Doing this provides more sensible diffs, rather than having to rewrap each time.

Indeed, the line lines were intentional.  Hopefully this will work out better than what we've been doing.  I suppose maybe this should be documented somewhere, though. :/

So, explicitly skipping that review point.  Others addressed above.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_19"></a>|@exarkun set status to `closed`|
|-|-|

(In [38574]) Merge http-getBody-6251-2

Author: tom.prince, exarkun
Reviewer: therve
Fixes: [#6251](https://github.com/twisted/twisted/issues/6251)

Introduce a Twisted Web client helper, `twisted.web.client.readBody`, for
extracting the body of an HTTP response all at once.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_20"></a>|@tomprince commented|
|-|-|

I filed the line length issue as [#6537](https://github.com/twisted/twisted/issues/6537), which also appears at [TheWayWeDoThingsNow#ideas](https://twistedmatrix.com/trac/wiki/TheWayWeDoThingsNow#ideas)
